### PR TITLE
feat: v2 - add store, world and schema-type, cli table code generation

### PR DESCRIPTION
**Pull request by holic**
This branch was used for fast iteration on `v2` packages. However, maintaining two separate branches (`main` and `v2`) causes unnecessary overhead, and the `v2` changes are mostly self-contained, so they can live side by side with the `v1` packages on `main`.

### CODE REVIEW

File: packages/world/test/World.t.sol [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+    WorldTestSystem system = new WorldTestSystem();
+    world.registerSystem("/testRoute", "/testSystem", system, false);
+
+    // Call the root sysyem
+    vm.expectEmit(true, true, true, true);
+    emit WorldTestSystemLog("call");
+    world.call("/testRoute/testSystem", abi.encodeWithSelector(WorldTestSystem.emitCallType.selector));
+  }
+
+  // TODO: add a test for systems writing to tables via the World
```
Comment (alvrs) [inline packages/world/test/World.t.sol:539]: link TODO with issue

File: packages/world/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,102 @@
+{
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig.json to read more about this file */
```
Comment (alvrs) [inline packages/world/tsconfig.json:3]: remove comments for readability

File: .github/workflows/gasreport.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        with:
+          version: nightly
+
+      - name: Install dependencies
+        run: yarn install --network-concurrency 1
+
+      - name: Install local CLI
+        run: cd packages/cli && yarn build && npm link && cd ../..
+
+      - name: Verify that gas report is up to date
```
[the remark this answers is not in the document]
Comment (alvrs) [reply .github/workflows/gasreport.yml:33]: https://github.com/latticexyz/mud/issues/442

File: packages/cli/src/render-table/field.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      _typedTableId,
+      _typedKeyArgs,
+      _typedFieldName,
+    ])}) internal {
+      ${_primaryKeysDefinition}
+      StoreSwitch.setField(_tableId, _primaryKeys, ${index}, ${renderEncodeField(field)});
+    }
+    `;
+
+    // TODO: this is super inefficient right now, need to add support for pushing to arrays to the store core library to avoid reading/writing the entire array
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/cli/src/render-table/field.ts:50]: added to https://github.com/latticexyz/mud/issues/438

File: packages/cli/src/render-table/record.ts [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  /** Set the full data using individual values */
+  function set(${renderArguments([
+    _typedTableId,
+    _typedKeyArgs,
+    renderArguments(options.fields.map(({ name, typeWithLocation }) => `${typeWithLocation} ${name}`)),
+  ])}) internal {
+    ${renderEncodedLengths(dynamicFields)}
+    bytes memory _data = abi.encodePacked(${renderArguments([
+      renderArguments(staticFields.map(({ name }) => name)),
+      // TODO try gas optimization (preallocate for all, encodePacked statics, and direct encode dynamics)
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/cli/src/render-table/record.ts:45]: added to https://github.com/latticexyz/mud/issues/444

File: packages/cli/src/utils/errors.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+export class NotESMConfigError extends Error {
+  name = "NotESMConfigError";
+  message = "MUD config must be an ES module";
+}
+
+export function logError(error: Error) {
+  if (error instanceof ValidationError) {
+    console.log(chalk.redBright(error.message));
+  } else if (error instanceof ZodError) {
+    // TODO currently this error shouldn't happen, use `fromZodErrorCustom`
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/cli/src/utils/errors.ts:28]: added to https://github.com/latticexyz/mud/issues/438

File: packages/cli/src/utils/errors.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    const validationError = fromZodError(error, {
+      prefixSeparator: "\n- ",
+      issueSeparator: "\n- ",
+    });
+    console.log(chalk.redBright(validationError.message));
+  } else if (error instanceof NotInsideProjectError) {
+    console.log(chalk.red(error.message));
+    console.log("");
+    console.log(chalk.blue(`To learn more about MUD's configuration, please go to [TODO link to docs]`));
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/cli/src/utils/errors.ts:37]: updated the link to https://mud.dev/packages/cli/ for now and added an issue to add docs for this here: https://github.com/latticexyz/mud/issues/445

File: packages/cli/src/utils/exec.ts [hunk 8]
```diff
@@ -1,40 +1,17 @@
...
   console.log("Cmd:");
   console.log([command, ...options].join(" "));
-  const [resolve, , promise] = deferred<{ exitCode: number; child: ChildProcess }>();
-
-  const child = spawn(command, options, { stdio: [process.stdin, process.stdout, process.stderr] });
 
-  child.on("exit", (code) => resolve({ exitCode: code ?? 0, child }));
-
-  return promise;
+  // TODO piping outputs and doing custom logging would be better for readability
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/cli/src/utils/exec.ts:13]: https://github.com/latticexyz/mud/issues/446

File: packages/schema-type/src/solidity/SchemaType.sol [hunk 9]
```diff
@@ -0,0 +1,237 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+using { getStaticByteLength } for SchemaType global;
+
+// WARNING: All enum changes MUST be mirrored for the other languages!
+// WARNING: SchemaType methods use hardcoded enum indexes, review them after any changes to the enum
+// TODO add and implement BYTES_ARRAY, STRING_ARRAY if they are needed (they are special nested arrays)
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/schema-type/src/solidity/SchemaType.sol:8]: https://github.com/latticexyz/mud/issues/447

File: packages/schema-type/test/solidity/SchemaType.t.sol [hunk 10]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import { Test } from "forge-std/Test.sol";
+import { SchemaType } from "../../src/solidity/SchemaType.sol";
+
+contract SchemaTypeTest is Test {
+  function testGetStaticByteLength() public {
+    assertEq(SchemaType.UINT8.getStaticByteLength(), 1);
+    // TODO add more tests
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/schema-type/test/solidity/SchemaType.t.sol:10]: added to https://github.com/latticexyz/mud/issues/444

File: packages/store/src/IStore.sol [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+  // Get partial data at schema index
+  function getField(
+    uint256 table,
+    bytes32[] calldata key,
+    uint8 schemaIndex
+  ) external view returns (bytes memory data);
+
+  // If this function exists on the contract, it is a store
+  // TODO: benchmark this vs. using a known storage slot to determine whether a contract is a Store
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/IStore.sol:55]: added to https://github.com/latticexyz/mud/issues/444

File: packages/store/src/Storage.sol [hunk 12]
```diff
@@ -0,0 +1,230 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import { console } from "forge-std/console.sol";
+import { Utils } from "./Utils.sol";
+
+/**
+ * TODO Probably not fully optimized
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/Storage.sol:8]: added to #444

File: packages/store/src/Storage.sol [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+  /**
+   * @notice Load raw bytes from storage at the given storagePointer, offset, and length
+   */
+  function load(
+    uint256 storagePointer,
+    uint256 length,
+    uint256 offset
+  ) internal view returns (bytes memory result) {
+    // TODO this will probably use less gas via manual memory allocation
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/Storage.sol:137]: added to #444

File: packages/store/src/StoreCore.sol [hunk 14]
```diff
@@ -0,0 +1,561 @@
...
+  error StoreCore_TableAlreadyExists(uint256 table);
+  error StoreCore_TableNotFound(uint256 table);
+  error StoreCore_NotImplemented();
+  error StoreCore_InvalidDataLength(uint256 expected, uint256 received);
+  error StoreCore_NoDynamicField();
+
+  /**
+   * Initialize internal tables.
+   * Consumers must call this function in their constructor.
+   * TODO: should we turn the schema table into a "proper table" and register it here?
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/StoreCore.sol:29]: added to #444

File: packages/store/src/Types.sol [hunk 15]
```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+using { getStaticByteLength } for SchemaType global;
+
+// WARNING: SchemaType methods use hardcoded enum indexes, review them after any changes to the enum
+// TODO add and implement BYTES_ARRAY, STRING_ARRAY if they are needed (they are special nested arrays)
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/Types.sol:7]: Added to #444

File: packages/store/src/tightcoder/TightCoder.sol [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+import { Slice, SliceLib } from "../Slice.sol";
+import { SchemaType } from "../Types.sol";
+
+library TightCoder {
+  /**
+   * @dev Copies the array to the location of `packedSlice`,
+   * tightly packing it using the given size per element (in bytes)
+   *
+   * TODO this function is currently not used externally and will be changed in the future
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/tightcoder/TightCoder.sol:12]: Added to #444

File: packages/store/src/tightcoder/TightCoder.sol [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    bytes32[] memory array,
+    Slice packedSlice,
+    uint256 elementSize,
+    bool leftAligned
+  ) private pure {
+    uint256 arrayLength = array.length;
+    uint256 packedPointer = packedSlice.pointer();
+    uint256 shiftLeft = leftAligned ? 0 : 256 - elementSize * 8;
+
+    // TODO temporary check to catch bugs, either remove it or use a custom error
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/tightcoder/TightCoder.sol:24]: Added to #444

File: packages/store/src/tightcoder/TightCoder.sol [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    uint256 packedPointer = packedSlice.pointer();
+    uint256 packedLength = packedSlice.length();
+    uint256 padLeft = leftAligned ? 0 : 256 - elementSize * 8;
+    // Array length (number of elements)
+    uint256 arrayLength;
+    unchecked {
+      arrayLength = packedLength / elementSize;
+    }
+
+    // TODO temporary check to catch bugs, either remove it or use a custom error
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/tightcoder/TightCoder.sol:80]: Added to #444

File: packages/world/src/World.sol [hunk 19]
```diff
@@ -0,0 +1,426 @@
...
+
+    // Register the hook
+    StoreCore.registerStoreHook(tableId, hook);
+  }
+
+  /**
+   * Register a hook for a given system route
+   */
+  function registerSystemHook(string calldata systemRoute, ISystemHook hook) public {
+    // TODO
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/world/src/World.sol:124]: Added to #444

File: packages/world/test/World.t.sol (hunk 1 above)
Comment (alvrs) [reply packages/world/test/World.t.sol:539]: Added to #444

**Review by alvrs — APPROVED** [review]

File: packages/cli/src/render-table/record.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/cli/src/render-table/record.ts:45]: moved to https://github.com/latticexyz/mud/issues/1132

File: packages/store/src/tightcoder/TightCoder.sol (hunk 16 above)
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/src/tightcoder/TightCoder.sol:12]: moved to https://github.com/latticexyz/mud/issues/1132

File: packages/store/src/tightcoder/TightCoder.sol (hunk 17 above)
[the remark this answers is not in the document]
Comment (dk1a) [reply packages/store/src/tightcoder/TightCoder.sol:25]: moved to https://github.com/latticexyz/mud/issues/1132

**Pull request merged**